### PR TITLE
EC Pro 2 PCB

### PR DESCRIPTION
**Pull request by Cipulot**
## Description

Addition of the EC Pro 2 PCB firmware. The EC Pro 2 is a replacement PCB kit for HHKB Professional 2 ANSI keyboards.
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/cipulot/ec_pro2/config.h
Comment (fauxpark) [inline keyboards/cipulot/ec_pro2/config.h:20]:
```suggestion
```

File: keyboards/cipulot/ec_pro2/matrix.c
Comment (fauxpark) [inline keyboards/cipulot/ec_pro2/matrix.c:19]:
```suggestion
```

File: keyboards/cipulot/ec_pro2/ec_pro2.c
Comment (fauxpark) [inline keyboards/cipulot/ec_pro2/ec_pro2.c:21]:
```suggestion
#include "ec_switch_matrix.h"
#include "quantum.h"
```

File: keyboards/cipulot/ec_pro2/ec_pro2.c
Comment (fauxpark) [inline keyboards/cipulot/ec_pro2/ec_pro2.c:159]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/cipulot/ec_pro2/readme.md
Comment (keyboard-magpie) [inline keyboards/cipulot/ec_pro2/readme.md:19]: Can you add some instructions here on entering bootloader mode, as per the [template](https://github.com/qmk/qmk_firmware/blob/master/data/templates/keyboard/readme.md). 

Otherwise looks good to me.

**Review by keyboard-magpie — CHANGES_REQUESTED** [review]

File: keyboards/cipulot/ec_pro2/readme.md
Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/readme.md:19]: Done 👍

**Review by keyboard-magpie — APPROVED** [review]

File: keyboards/cipulot/ec_pro2/ec_pro2.c
Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/ec_pro2.c:21]: @fauxpark regarding this change, I tested the suggested changes for the `include` and those are working.

About the `#ifdef VIA_ENABLE` and relative `#endif`, I implemented that so that the following functionality is only enabled when VIA is enabled. In case someone compiles the code with `default` as target keymap the default  actuation levels used are assigned in the `matrix.c` file. In here those will be overwritten with the ones in eeprom if VIA enabled.

I'll push for now the changes for the `includes` and wait to apply the removal of the `#ifdef` till your response on this 👍

Comment (fauxpark) [reply keyboards/cipulot/ec_pro2/ec_pro2.c]: If this code is only used when VIA is enabled, then it should not be unconditionally added to `SRC`.

Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/ec_pro2.c]: The addition of the `apc.c` file without the condition tho generates errors if I compile the `default` keymap, since it tries to reference variables that are only defined when the VIA stuff is imported.

File: keyboards/cipulot/ec_pro2/ec_switch_matrix.c
Comment (sigprof) [inline keyboards/cipulot/ec_pro2/ec_switch_matrix.c:120]: Technically this is wrong, because `ATOMIC_BLOCK_FORCEON` is `chSysLock(); ...; chSysUnlock();`, and `adcConvert()` has its own `osalSysLock()`/`osalSysUnlock()` pair inside.  This code will panic if `CH_DBG_SYSTEM_STATE_CHECK` is enabled; it just happens to work without debug checks, because the duplicate lock and unlock calls are ignored.

Doing this properly may require either some new analog API, or using the ChibiOS ADC driver directly (`adcStartConversionI()` can be called from the S-Locked state, but waiting for the conversion completion would require working interrupts, so it can't be done without leaving the S-Locked state temporarily — either explicitly, or through a thread suspend, which is what `adcConvert()` does internally).

Also I suppose that what you want here is to avoid interrupts messing up timings — but in that case the atomic block should also cover at least `charge_capacitor(row);` instead of having a separate atomic block inside that function (otherwise an interrupt can be served between those atomic blocks).

File: keyboards/cipulot/ec_pro2/ec_pro2.c
Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/ec_pro2.c]: I'll revert the latest change since it breaks the `default` keymap compile.

Comment (fauxpark) [reply keyboards/cipulot/ec_pro2/ec_pro2.c]: You need to move it to the via keymap's rules.mk. That's what I'm referring to by "unconditionally added to SRC."

File: keyboards/cipulot/ec_pro2/ec_switch_matrix.c
Comment (sigprof) [inline keyboards/cipulot/ec_pro2/ec_switch_matrix.c:42]: These atomic blocks should probably be removed (the one inside `discharge_capacitor()` does not have anything to protect against, and the one inside `charge_capacitor()` should become a single atomic block that covers both the pin manipulations and the ADC conversion start).

**Review by sigprof — CHANGES_REQUESTED** [review]

File: keyboards/cipulot/ec_pro2/ec_pro2.c
Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/ec_pro2.c]: Ohhhhh, ok, I'm doodo. I'll do it now

File: keyboards/cipulot/ec_pro2/ec_switch_matrix.c
Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/ec_switch_matrix.c:120]: I see so the options are to wrap the `charge_capacitor(row)`, `sw_value = adc_read(adcMux)` and `discharge_capacitor()` in a single `ATOMIC_BLOCK_FORCEON` and remove the ones in the functions or remove it all together. 

I tested the first option and the board behaves as expected. If that's an ok solution for you @sigprof I'll push a commit for that.

The change basically looks like this:
```c
ATOMIC_BLOCK_FORCEON {
        // Set the row pin to high state and have capacitor charge
        charge_capacitor(row);

        ATOMIC_BLOCK_FORCEON {
            sw_value = adc_read(adcMux);
        }
        // Discharge peak hold capacitor
        discharge_capacitor();
    }
```
I'll wait for your confirmation to push this.

File: keyboards/cipulot/ec_pro2/ec_pro2.c
Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/ec_pro2.c]: @fauxpark I've moved the entire functionality inside of the VIA keymap folder, removing the `ec_pro2.c` file and adding the new file containing the functions, `via_apc.c` in the `SRC` of VIA's `rules.mk`. 

Changes tested and working. Now both the via and default keymap targets compile without problems.

File: keyboards/cipulot/ec_pro2/ec_switch_matrix.c
Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/ec_switch_matrix.c:42]: Covered in the prev request change.

File: keyboards/cipulot/ec_pro2/keymaps/via/via_apc.c
Comment (fauxpark) [inline keyboards/cipulot/ec_pro2/keymaps/via/via_apc.c:19]:
```suggestion
#include "ec_switch_matrix.h"
#include "eeconfig.h"
````

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/cipulot/ec_pro2/ec_switch_matrix.c
Comment (sigprof) [reply keyboards/cipulot/ec_pro2/ec_switch_matrix.c]: Wrapping `discharge_capacitor()` into the atomic block is probably not needed — `adc_read()` will reenable interrupts anyway while it's waiting for the ADC conversion to finish, and some other interrupt can happen during that time.  Does an extra delay before `discharge_capacitor()` matter here?

As for making the code actually correct (and not just “works with debug checks disabled”), I'm not sure what could be done here.  It could be possible to add something like `adc_read_atomic()` that is safe to call inside an atomic block (by replacing the `adcConvert()` call in the existing implementation with a copy of its internal implementation with lock/unlock removed); but it won't be really atomic, because waiting for the conversion completion needs to be done with interrupts enabled.  Also there would be some issues with the ADC initialization (`adc_read()` does that lazily, but calling `adcStart()` from the S-Locked state is not allowed).

Comment (sigprof) [reply keyboards/cipulot/ec_pro2/ec_switch_matrix.c]: BTW, your current code also has that ADC init issue; you may perform a dummy `adc_read()` in the init code to make sure that `adcStart()` has been called in the appropriate state.

File: keyboards/cipulot/ec_pro2/keymaps/via/via_apc.c
Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/keymaps/via/via_apc.c]: Actually, this causes errors at compile time with the `via_command_ids`. 
Adding `#include "eeconfig.h"` fixes the via ids but prompts unknown type names for keycodes in via.h compile. Reverting back to `quantum.h` to fix these errors.

File: keyboards/cipulot/ec_pro2/keymaps/via/via_apc.c
Comment (fauxpark) [inline keyboards/cipulot/ec_pro2/keymaps/via/via_apc.c:18]:
```suggestion
#include "eeconfig.h"
#include "via.h"
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/cipulot/ec_pro2/keymaps/via/via_apc.c
Comment (fauxpark) [reply keyboards/cipulot/ec_pro2/keymaps/via/via_apc.c]: quantum.h is overkill here, you should only include the headers that provide symbols you're using.

Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/keymaps/via/via_apc.c]: Screenshot 2023-02-23 at 4 08 39 PM

Indeed that's overkill. Including the following gives the error in the picture.

File: keyboards/cipulot/ec_pro2/ec_switch_matrix.c
Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/ec_switch_matrix.c]: Having some interrupt delaying the `discharge_capacitor()` should really matter, I can definitely leave that one out especially since the `adc_read()` reenables interrupts as you said. I tested that and it works as intended. 
I'm inclined to use then the following and "leave it there" as tested working for now, until I come up with something more coherent in the actual ADC implementations.

```c
ATOMIC_BLOCK_FORCEON {
        // Set the row pin to high state and have capacitor charge
        charge_capacitor(row);
        // Read the ADC value
        sw_value = adc_read(adcMux);
    }
// Discharge peak hold capacitor
discharge_capacitor();
```

Regarding the dummy `adc_read()`, I was thinking about calling it in my `ecsm_init` function right after the `adcMux = pinToMux(ANALOG_PORT);`. 

If that's ok I'll commit these and push.

File: keyboards/cipulot/ec_pro2/keymaps/via/via_apc.c
Comment (fauxpark) [reply keyboards/cipulot/ec_pro2/keymaps/via/via_apc.c]: This shouldn't be an issue in a few days, the keycodes have been removed from via.h in develop.

Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/keymaps/via/via_apc.c]: I pushed a fix for that error. If it needs to be revised when the new merge happens I'll handle it

File: keyboards/cipulot/ec_pro2/ec_switch_matrix.c
Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/ec_switch_matrix.c]: Screenshot 2023-02-23 at 5 12 28 PM

Screenshot 2023-02-23 at 5 12 46 PM

File: keyboards/cipulot/ec_pro2/readme.md
Comment (drashna) [inline keyboards/cipulot/ec_pro2/readme.md:3]: This uses a smaller, but still high res image that is more suited for github's markdown rendering.
```suggestion
![EC Pro 2 PCB](https://i.imgur.com/uYOgTYoh.png)
```

File: keyboards/cipulot/ec_pro2/info.json
Comment (fauxpark) [inline keyboards/cipulot/ec_pro2/info.json:14]:
```suggestion
```

File: keyboards/cipulot/ec_pro2/config.h
Comment (fauxpark) [inline keyboards/cipulot/ec_pro2/config.h:19]:
```suggestion
#define MATRIX_ROWS 5
#define MATRIX_COLS 15

/* Custom matrix pins and port select array */
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/cipulot/ec_pro2/readme.md
Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/readme.md]: Done 👍

File: keyboards/cipulot/ec_pro2/config.h
Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/config.h:22]: Done in batch commit for removal in `info.json` and addition to `config.h`👍

**Comment by sigprof** [comment]
The layout in `info.json` looks wrong:
layout on config.qmk.fm
If the PCB is intended to support multiple layouts, you should have `LAYOUT_all` with all possible key positions (it may need to have something unusual in the Enter area, like a 1.25u Enter), and `LAYOUT_60_hhkb` and `LAYOUT_60_iso_hhkb` as some layouts that could actually be built (`60_hhkb` is a known community layout and needs to be declared as such, `60_iso_hhkb` is something non-standard).

**Comment by Cipulot** [comment]
@sigprof Will do the `LAYOUT_all` and the `60_hhkb`. For the other configs I'll just count on the `LAYOUT_all` when using VIA down the line. This also in anticipation of the other PR that I'll open once this gets merged for a much more complex set of layout options. Realistically the basic ansi option is good for the configurator.

**Comment by Cipulot** [comment]
- @sigprof changes to fix the faulty `info.json` and expand the use of default layouts implemented as requested and tested.
- @fauxpark after applying the latest Breaking Changes I've removed the unnecessary includes from the custom code file.

If these are ok I'll use the same suggestions to prepare followup PRs for the rest of the EC series once this gets final approval and merge.

File: keyboards/cipulot/ec_pro2/info.json
Comment (fauxpark) [inline keyboards/cipulot/ec_pro2/info.json:55]: `"w": 1` is unnecessary, the default key width and height are 1u.

Comment (Cipulot) [reply keyboards/cipulot/ec_pro2/info.json]: @fauxpark done 👍

**Comment by Cipulot** [comment]
Waiting for feedback on the `LAYOUT` stuff and the rest of the solved requests so that I can patch the other boards before submitting the rest of the lineup for PR.

**Comment by Cipulot** [comment]
@sigprof @sigprof friendly ping 😊

**Comment by Cipulot** [comment]
The latest commit reflects changes brought up by zvecr in regard to the use of `_kb` eeprom functions instead of `_user` at a keymap level code. 
It also removed some items from `info.json` that aren't supposed to be used (keyboard folder path) or are the default choice(board).
Changes have already been tested on hardware and confirmed to be working. This PR now follows the same pattern as the other recently opened by me for the rest of the EC boards.

**Review by drashna — APPROVED** [review]

**Pull request merged**
